File: 8-string-to-integer-atoi/string-to-integer-atoi.py
class Solution:
    def myAtoi(self, s: str) -> int:
        s = s.lstrip()
        if not s: return 0
        sign, i, res = -1 if s[0] == '-' else 1, 1 if s[0] in '+-' else 0, 0
        while i < len(s) and s[i].isdigit():
            res = res * 10 + int(s[i])
            i += 1
        res *= sign
        return max(-2**31, min(res, 2**31 - 1))

File: 8-string-to-integer-atoi/string-to-integer-atoi.cpp
class Solution {
public:
    int myAtoi(string s) {
        int i = 0, n = s.length();
        long ans = 0;
        int sign = 1;

        // 1. Skip leading whitespaces
        while (i < n && s[i] == ' ') i++;

        // 2. Check for sign
        if (i < n && (s[i] == '+' || s[i] == '-')) {
            sign = (s[i] == '-') ? -1 : 1;
            i++;
        }

        // 3. Parse digits
        while (i < n && isdigit(s[i])) {
            ans = ans * 10 + (s[i] - '0');
            
            // 4. Clamp if out of bounds
            if (sign * ans <= INT_MIN) return INT_MIN;
            if (sign * ans >= INT_MAX) return INT_MAX;

            i++;
        }

        return sign * ans;
    }
};
